Update react and react-hooks

oleggrishechkin opened the pull request:
it PR slightly increase react-hooks performance (useCallback is missing in current implementation), and rewrite react with classes - now react-hooks and react implementations are same - first use memo, useReducer and useCallback, second use PureComponent, state and methods (one dispatch instead of one method per action)

react-hooks performance are great, and react with classes so slow. hooks:classes - 1:0

ryansolid commented:
Hmm.. a lot of changes here. And the React classes implementation was definitely in need of an overhaul. On the use of useCallback there was this previous PR #831 . Im concerned with all the changes whether we can no for sure in isolation which way is actually faster. The original implementation had them and then they were intentional removed.

oleggrishechkin commented:
@ryansolid 
I know how improve performance without useCallback - all row data (id, label, etc) should be moved in othe memoized component like RowData, but i think it's cheating too.

I post results in [issue](https://github.com/krausest/js-framework-benchmark/issues/838) with optimized react implementation with mutations, global state and direct subscribe to updated selection, label and data array. It's close to react-rxjs, but i get not so good results for react-rxjs on my machine

ryansolid commented on frameworks/keyed/react-hooks/src/main.jsx:
```diff
@@ -34,35 +39,47 @@ function listReducer(state, action) {
     case 'SWAP_ROWS':
-      debugger;
-      return (data.length > 998) ? { data: [data[0], data[998], ...data.slice(2, 998), data[1], data[999]] } : state;
+      return { data: [data[0], data[998], ...data.slice(2, 998), data[1], data[999]] };
```
Probably shouldn't remove the length check here. While unnecessary for the test runs it is necessary for the operation not to die horribly after a couple of rows have been removed.

ryansolid commented:
@oleggrishechkin Yeah looking closer the Hooks differences are actually probably mostly Prettier. It's the removal of the length check on SWAP_ROWS.. the removal of the return `true` function on memo (I'm gathering it is unnecessary?) and the re-introduction of the useCallback, hoisting initial state, and some stylish stuff to use more destructuring and less property access. Am I missing anything? I'm just unclear where the perf gain is coming from. Because looks like mostly just undoes the last PR which was shown to be faster?

To be fair I never tested that myself, and I wrote the original hooks implementation with `useCallback`. I understand what you mean through different structure we could remove any need for `useCallback`. I'm just trying to understand this sort of back and forth tug of war with `useCallback`. Is there some subtle difference here that makes it different than it was before #831 in terms of performance?

oleggrishechkin commented:
@ryansolid, i think i concentrate on select row performance when say "performace increased"

Yer, mostly differences is Prettier and react with classes refactor - i use PureComponent, remove forceUpdate, create dispatch method like in hooks.

I added useCallback, because it was imported, but not used, and i think that somebody want add useCallback but forget about it. Also it's good practice to prevent unnecessary re-rendering - in classes you should use methods, in hooks - useCallback. But in this benchmark useCallbacks in each row is decrease rows creating performance, but in Real World example you prefer use virtualization/windowing instead of create so much rows

oleggrishechkin commented:
@ryansolid i added length check for swap, removed `useCallbacks`, and added `shouldComponentUpdate`, also i added second parameter to `memo` in react-hooks, it's should exclude `dispatch` while checking props eq and (i beat react-rxjs with it in my tests)

ryansolid commented:
Nice. Yeah I'm not here to state I know which way is faster. I will take whatever averages out faster. I just was unclear what that is when presented with seemingly contradictory information. If you test both and one churns out better scores I think we should go for it.

oleggrishechkin commented:
I did some tests across react-redux-hooks, react-hooks, react-rxjs (fastest in react case). I tested "with `useCallback`", with `memo` second parameter, with components for row label, row remove cross, with "`useSelector` on each row" and without. 

Fastest in all cases except select row - with memo second parameter (for react-hooks), pure row function component (without `useSelector` and `useCallback`). Other combinations was slower. So, that's why react-rxjs implementation fastest for now - it's use global states, and you no need provide dispatch prop.

In this PR react-hooks performance should be closer to react-rxjs

krausest commented:
Thanks. Here's a screenshot from the result (left new, right old). Maybe it's a little bit faster for create row and select row.
<img width="1013" src="https://user-images.githubusercontent.com/3374055/112832783-195dbd80-9096-11eb-8d88-4da2ab16c453.png" />
